File: SunkiojiDalis/src/Helpers/Vector2D.ts
export class Vector2D{
    x: number;
    y: number;

    constructor(x: number, y: number){
        this.x = x;
        this.y = y;
    }

    GetMagnidute(): number {
        return Math.sqrt(this.x ** 2 + this.y ** 2);
    }

    Normalize(): Vector2D {
        let magnitude = this.GetMagnidute();
        return new Vector2D(this.x / magnitude, this.y / magnitude);
    }

    DirectionTo(to: Vector2D): Vector2D {
        return Vector2D.Subtract(to, this);
    }
    
    static Lerp(origin: Vector2D, target: Vector2D, t: number): Vector2D {
        let direction = origin.DirectionTo(target);
        return new Vector2D(origin.x + direction.x * t, origin.y + direction.y * t);
    }

    static ProjectOn(vector: Vector2D, prjectionVector: Vector2D): Vector2D {
        let normaizedV = prjectionVector.Normalize();
        let dotProduct = Vector2D.DotProduct(vector, normaizedV);
        return Vector2D.Multiply(normaizedV, dotProduct);
    }

    static DotProduct(v1: Vector2D, v2: Vector2D): number {
        return v1.x * v2.x + v1.y * v2.y;
    }

    static Equals(v1: Vector2D, v2: Vector2D): boolean {
        return v1.x === v2.x && v1.y === v2.y;
    }

    static Add(v1: Vector2D, v2: Vector2D): Vector2D{
        return new Vector2D(v1.x + v2.x, v1.y + v2.y);
    }

    static Subtract(v1: Vector2D, v2: Vector2D){
        return new Vector2D(v1.x - v2.x, v1.y - v2.y);
    }

    static Multiply(v: Vector2D, num: number){
        return new Vector2D(v.x * num, v.y * num);
    }
    
    static Divide(v: Vector2D, num: number){
        return new Vector2D(v.x / num, v.y / num);
    }
}

File: SunkiojiDalis/Hubs/Helpers/Vector2D.cs
using System;
public class Vector2D
{
    public float X { get; set; } = 0;
    public float Y { get; set; } = 0;
    public Vector2D(float x, float y)
    {
        this.X = x;
        this.Y = y;
    }

    public float GetMagnidute() 
    {
        return (float)Math.Sqrt(Math.Pow(this.X, 2) + Math.Pow(this.Y, 2));
    }

    public Vector2D Normalize() 
    {
        float magnitude = this.GetMagnidute();
        return this / magnitude;
    }

    public Vector2D DirectionTo(Vector2D to) 
    {
        return to - this;
    }
    
    public static Vector2D Lerp(Vector2D origin, Vector2D target, float t)
    {
        Vector2D direction = origin.DirectionTo(target);
        return origin + direction * t;
    }

    public static Vector2D ProjectOn(Vector2D vector, Vector2D prjectionVector) 
    {
        Vector2D normalizedV = prjectionVector.Normalize();
        float dotProduct = Vector2D.DotProduct(vector, normalizedV);
        return normalizedV * dotProduct;
    }

    public static float DotProduct(Vector2D v1, Vector2D v2)
    {
        return v1.X * v2.X + v1.Y * v2.Y;
    }

    public override bool Equals(object obj)
    {
        return base.Equals(obj);
    }

    public override int GetHashCode()
    {
        return base.GetHashCode();
    }
#region Operators
    public static bool operator == (Vector2D v1, Vector2D v2)
    {
        if(v1 is null && v2 is null) return true;
        if(v1 is null) return false;
        if(v2 is null) return false;

        return v1.X == v2.X && v1.Y == v2.Y;
    }

    public static bool operator != (Vector2D v1, Vector2D v2)
    {
        if(v1 is null && v2 is null) return false;
        if(v1 is null) return true;
        if(v2 is null) return true;

        return v1.X != v2.X || v1.Y != v2.Y;
    }

    public static Vector2D operator + (Vector2D v1, Vector2D v2)
    {
        return new Vector2D(v1.X + v2.X, v1.Y + v2.Y);
    }

    public static Vector2D operator - (Vector2D v1, Vector2D v2)
    {
        return new Vector2D(v1.X - v2.X, v1.Y - v2.Y);
    }

    public static Vector2D operator * (Vector2D v, float num)
    {
        return new Vector2D(v.X * num, v.Y * num);
    }
    public static Vector2D operator * (float num, Vector2D v)
    {
        return new Vector2D(v.X * num, v.Y * num);
    }

    public static Vector2D operator / (Vector2D v, float num)
    {
        return new Vector2D(v.X / num, v.Y / num);
    }

    public static Vector2D operator / (float num, Vector2D v)
    {
        return new Vector2D(v.X / num, v.Y / num);
    }
#endregion
}